core/src/main/java/com/scalar/db/transaction/consensuscommit/RollbackMutationComposer.java

```diff
@@ -128,12 +118,31 @@ private Delete composeDelete(Operation base, TransactionResult result) {
         .withConsistency(Consistency.LINEARIZABLE);
   }
 
-  private Optional<TransactionResult> getLatestResult(Operation operation, TransactionResult result)
-      throws ExecutionException {
+  private Optional<TransactionResult> getLatestResult(
+      Operation operation, @Nullable TransactionResult result) throws ExecutionException {
+    Key partitionKey;
+    Key clusteringKey;
+    if (operation instanceof Mutation) {
+      // for usual rollback
+      partitionKey = operation.getPartitionKey();
+      clusteringKey = operation.getClusteringKey().orElse(null);
+    } else {
+      assert operation instanceof Selection;
+      if (result != null) {
+        // for rollback in lazy recovery
+        partitionKey = result.getPartitionKey().get();
+        clusteringKey = result.getClusteringKey().orElse(null);
+      } else {
+        // for deleting non-existing record that was prepared with DELETED for Serializable with
+        // Extra-write
+        assert operation instanceof Get;
+        partitionKey = operation.getPartitionKey();
+        clusteringKey = operation.getClusteringKey().orElse(null);
+      }
+    }
```

Comment:
This is the logic to get partition key and clustering key for `RollbackMutationComposer`.